src/main/java/de/hysky/skyblocker/skyblock/dwarven/CarpetHighlighter.java

```diff
@@ -0,0 +1,105 @@
+package de.hysky.skyblocker.skyblock.dwarven;
+
+import de.hysky.skyblocker.annotations.Init;
+import de.hysky.skyblocker.config.SkyblockerConfigManager;
+import de.hysky.skyblocker.events.SkyblockEvents;
+import de.hysky.skyblocker.utils.Boxes;
+import de.hysky.skyblocker.utils.Location;
+import de.hysky.skyblocker.utils.Resettable;
+import de.hysky.skyblocker.utils.render.RenderHelper;
+import de.hysky.skyblocker.utils.render.Renderable;
+import de.hysky.skyblocker.utils.scheduler.Scheduler;
+import it.unimi.dsi.fastutil.objects.ObjectAVLTreeSet;
+import net.fabricmc.fabric.api.client.networking.v1.ClientPlayConnectionEvents;
+import net.fabricmc.fabric.api.client.rendering.v1.WorldRenderContext;
+import net.fabricmc.fabric.api.client.rendering.v1.WorldRenderEvents;
+import net.minecraft.block.BlockState;
+import net.minecraft.block.Blocks;
+import net.minecraft.block.CarpetBlock;
+import net.minecraft.client.MinecraftClient;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Vec3d;
+
+import java.awt.*;
+
+/**
+ * Highlights unbreakable carpets within ore veins in the Dwarven Mines.
+ */
+public final class CarpetHighlighter implements Renderable, Resettable {
+	public static final CarpetHighlighter INSTANCE = new CarpetHighlighter();
+
+	private static final Vec3d CARPET_BOUNDING_BOX = Boxes.getLengthVec(CarpetBlock.SHAPE.getBoundingBox());
+	private static final int SEARCH_RADIUS = 15;
+	private static final int TICK_INTERVAL = 15;
+	private static final ObjectAVLTreeSet<BlockPos> CARPET_LOCATIONS = new ObjectAVLTreeSet<>();
+
+	private static float[] colorComponents;
+	private static int tickCounter = 0;
+	private static boolean isLocationValid = false;
+
+	@Init
+	public static void init() {
+		INSTANCE.configCallback(SkyblockerConfigManager.get().mining.dwarvenMines.carpetHighlightColor);
+		WorldRenderEvents.AFTER_TRANSLUCENT.register(INSTANCE::render);
+		SkyblockEvents.LOCATION_CHANGE.register(INSTANCE::onLocationChange);
+		Scheduler.INSTANCE.scheduleCyclic(INSTANCE::tick, TICK_INTERVAL);
+		ClientPlayConnectionEvents.JOIN.register(INSTANCE);
+	}
+
+	@Override
+	public void render(WorldRenderContext context) {
+		if (!isLocationValid || !SkyblockerConfigManager.get().mining.dwarvenMines.enableCarpetHighlighter) return;
+		for (BlockPos carpetLocation : CARPET_LOCATIONS) {
+			RenderHelper.renderFilled(context, carpetLocation, CARPET_BOUNDING_BOX, colorComponents, colorComponents[3], false);
+		}
+	}
+
+	public void onLocationChange(Location location) {
+		isLocationValid = location == Location.DWARVEN_MINES;
+	}
+
+	public void tick() {
+		if (!isLocationValid || MinecraftClient.getInstance().world == null || MinecraftClient.getInstance().player == null) return;
```

Comment:
This should also check for whether the feature is enabled or not so that we are not locating carpets unnecessarily.

Reply:
Done in [`fffa5bc`](https://github.com/SkyblockerMod/Skyblocker/pull/1034/commits/fffa5bc82428080578dc3e1149d181a74fd5f1f6).